### ISSUE TITLE
Self training seems to cost too much time

### ISSUE THREAD

**Issue by xiaoxin83121** [CVMI-Lab/ST3D#54]
Hello, thanks a lot for your clear and great code.
I have tried to reproduce the domain adaptation result from waymo to kitti. However, I found that it cost me nearly 3h/epoch with 4 3080. Should the self-training be as fast as training on naive OpenPCDet? Or the memory ensembling costs that much?
Hope to get your reply.

**Comment by jihanyang** [comment]
No. I takes only 5-10 minutes per epoch.

**Comment by xiaoxin83121** [comment]
> No. I take only 5-10 minutes per epoch.

Thanks for your reply. I'll let you know when I find out.

**Comment by xiaoxin83121** [comment]
Hi, I find that there are tensorflow and tensorflow-gpu in my environment. I uninstall the tensorflow-cpu which binded to waymo-dataset-tools and now it takes 15min per epoch. Thanks a lot for your information.

**Comment by xiaoxin83121** [comment]
> Hi, I find that there are tensorflow and tensorflow-gpu in my environment. I uninstall the tensorflow-cpu which binded to waymo-dataset-tools and now it takes 15min per epoch. Thanks a lot for your information.

Hi again, cpu/gpu version do influence the training time. However, after updating the pseudo in 2 epochs, training time comes to 2h again. And a RecursionError of python occurs.
The real reason is the **inferior pseudo label**. When useing function [_\_\_item\_\__](https://github.com/CVMI-Lab/ST3D/blob/master/pcdet/datasets/kitti/kitti_dataset.py#L415), it will do [_prepare_data_](https://github.com/CVMI-Lab/ST3D/blob/master/pcdet/datasets/dataset.py#L239) to filter some inferior bbox. And in function _prepare_data_, if len(bbox)==0, _\_\_item\_\__ will be used again in a recursion manner. 
I have loaded checkpoint in early stage of source-training with mistake. And the inferior pseudo bboxes significantly increase the data iteration time.

**Comment by jihanyang** [comment]
Yes. If you have unsuitable score thresh or inferior pre-trained model, this could happen.

**Comment by xiaoxin83121** [comment]
> Yes. If you have unsuitable score thresh or inferior pre-trained model, this could happen.

I have tried to train on source waymo dataset and self-train on target kitti dataset with the config yaml provided. But I got 
```
## Source_Only
Car AP@0.70, 0.70, 0.70:
bbox AP:88.3465, 82.6289, 83.2283
bev  AP:80.7325, 72.2854, 71.2048
3d   AP:38.3196, 35.7308, 35.8998
aos  AP:88.33, 82.44, 82.97

## ST3D
Car AP@0.70, 0.70, 0.70:
bbox AP:89.4433, 85.8820, 84.7858
bev  AP:82.1983, 75.4475, 74.1390
3d   AP:49.0421, 46.3101, 45.0632
aos  AP:89.40, 85.68, 84.47
```
which the result on source-only is better than that in paper(67.64 / 27.48) and the result of ST3D is much worse than (82.19 / 61.83)
Could you offer me the ros-pretrained model on waymo?

**Comment by jihanyang** [comment]
To access these pretrained models, please send us an email with your name, institute, a screenshot of the the Waymo dataset registration confirmation mail, and your intended usage. Please send a second email if we don't get back to you in two days. Please note that Waymo open dataset is under strict non-commercial license so we are not allowed to share the model with you if it will used for any profit-oriented activities.

**Comment by xiaoxin83121** [comment]
> To access these pretrained models, please send us an email with your name, institute, a screenshot of the the Waymo dataset registration confirmation mail, and your intended usage. Please send a second email if we don't get back to you in two days. Please note that Waymo open dataset is under strict non-commercial license so we are not allowed to share the model with you if it will used for any profit-oriented activities.

Thanks a lot for your help. I have sent an email to [jhyang@eee.hku.hk].
